### PR TITLE
feat(ethexe): Initial on-chain `db-sync` request validation

### PR DESCRIPTION
**Pull request by ark0f**
It's not yet possible to test actual validation with `alloy` until `db-sync` protocol is actually used in service

### CODE REVIEW

File: ethexe/network/src/db_sync/ongoing.rs [hunk 1]
```diff
@@ -113,25 +152,53 @@ impl OngoingRequest {
         response
     }
 
+    fn inner_complete(
+        mut self,
+        peer: PeerId,
+        response: Response,
+    ) -> Result<(RequestId, Response), Self> {
+        if let Some(new_request) = self.request.difference(&response) {
```
Comment (grishasobol) [inline ethexe/network/src/db_sync/ongoing.rs:160]: What is `validate_response` inside `difference` impl?

Comment (ark0f) [reply ethexe/network/src/db_sync/ongoing.rs:160]: Artifact after renamings

**Review by grishasobol — APPROVED** [review]

File: ethexe/cli/src/service.rs [hunk 2]
```diff
@@ -53,6 +53,7 @@ pub struct Service {
     db: Database,
     observer: ethexe_observer::Observer,
     query: ethexe_observer::Query,
+    router_query: RouterQuery,
```
Comment (breathx) [inline ethexe/cli/src/service.rs:56]: can't you create router query from observer?

File: ethexe/cli/src/service.rs [hunk 3]
```diff
@@ -781,6 +800,29 @@ impl Service {
...
+        let response = validating_response.response();
+
+        if let db_sync::Response::ProgramIds(ids) = response {
+            let ethereum_programs = router_query.programs_count().await?;
+            if ethereum_programs != U256::from(ids.len()) {
+                return Ok(false);
+            }
+
+            for &id in ids {
+                let code_id = router_query.program_code_id(id).await?;
```
Comment (breathx) [inline ethexe/cli/src/service.rs:817]: wdyt about at least putting issue for multiple queries? like by batches of 32 ids mb?

**Review by breathx — APPROVED** [review]

File: ethexe/cli/src/service.rs (hunk 2 above)
Comment (ark0f) [reply ethexe/cli/src/service.rs:56]: I can but what's the point? It is already created in `Service::new()`

**Pull request merged**